internal/monitors/device-updater/device_updater.go

```diff
@@ -47,42 +50,119 @@ func (d *DeviceUpdater) UpdateDevices() {
+					log.Errorf("failed to update device %s (fleet %s): %v", *dev.Metadata.Name, fleets[fleetIndex].Name, err)
 				}
 			}
 		}
+
+		// Handle devices that don't match the fleet's label selector but still have the owner
+		err = d.handleOwningFleetOrphaned(ctx, log, &fleets[fleetIndex])
+		if err != nil {
+			log.Errorf("failed to update orphaned devices for fleet %s: %v", fleets[fleetIndex].Name, err)
+		}
+	}
+}
+
+/*
+Handle the case where the device matches a fleet that isn't its owner. This
+can happen if the selector changed or the device's labels changed. We need to
+check that the current owner fleet doesn't still match the device before we
+update the device's owner, because that means the descriptors of two fleets
+match the device.
+*/
+func (d *DeviceUpdater) handleOwningFleetChanged(ctx context.Context, log logrus.FieldLogger, device *api.Device, fleet *model.Fleet) error {
+	// "fleet" is potentially the new owner of "device" because, but we first need
+	// to make sure that the label selectors of both the current fleet and the new
+	// fleet aren't a match for this device.
+	currentOwningFleet, err := d.fleetStore.Get(ctx, fleet.Resource.OrgID, *device.Metadata.Owner)
+	if err != nil && err != gorm.ErrRecordNotFound {
+		return err
+	}
+	if currentOwningFleet != nil && util.LabelsMatchLabelSelector(*device.Metadata.Labels, currentOwningFleet.Spec.Selector.MatchLabels) {
+		d.log.Warnf("Device %s matches more than one fleet (%s and %s)", *device.Metadata.Name, *device.Metadata.Owner, fleet.Name)
+		// Oh no, a device with 2 owners! How do we indicate that?
+	} else {
+		log.Infof("Updating fleet of device %s from %s to %s", *device.Metadata.Name, *device.Metadata.Owner, fleet.Name)
+		device.Metadata.Owner = util.SetResourceOwner(model.FleetKind, fleet.Name)
+	}
+
+	return nil
+}
+
+/*
+Handle devices that don't match their owner fleet's selector.
+*/
+func (d *DeviceUpdater) handleOwningFleetOrphaned(ctx context.Context, log logrus.FieldLogger, fleet *model.Fleet) error {
+	// Perform a query of devices that don't match the label selector but still have the owner
+	listParams := service.ListParams{
+		Labels:       fleet.Spec.Data.Selector.MatchLabels,
+		InvertLabels: util.BoolToPtr(true),
+		Owner:        util.SetResourceOwner(model.FleetKind, fleet.Name),
+		Limit:        service.MaxRecordsPerListRequest, //TODO: Handle paging
+	}
+	devices, err := d.devStore.List(ctx, fleet.OrgID, listParams)
+	if err != nil {
+		return fmt.Errorf("failed to list devices that no longer belong to fleet %s: %v", fleet.Name, err)
+	}
+
+	for devIndex := range devices.Items {
+		dev := devices.Items[devIndex]
+		d.log.Infof("Unsetting fleet for device %s (was %s)", *dev.Metadata.Name, *dev.Metadata.Owner)
+		dev.Metadata.Owner = util.StrToPtr("")
+		_, _, err := d.devStore.CreateOrUpdate(ctx, fleet.OrgID, &dev)
+		if err != nil {
+			return fmt.Errorf("failed to update device without fleet %s: %v", *dev.Metadata.Name, err)
```

Comment:
wrapping errors can help with better testing and fine grained error control.
```suggestion
			return fmt.Errorf("failed to update device without fleet %s: %w", *dev.Metadata.Name, err)
```